float32 val
---
bool success